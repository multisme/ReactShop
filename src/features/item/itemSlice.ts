import { createSlice } from "@reduxjs/toolkit";
import { AppThunk } from "../../app/store";

export interface itemData {
  id: number | undefined;
  quantity: number;
  name: string;
  price?: number;
  details?: string;
  url?: string;
}

export const initialState = {
  id: undefined,
  quantity: 0,
  name: "unknown",
} as itemData;

const ItemSlice = createSlice({
  name: "item",
  initialState,
  reducers: {},
});
