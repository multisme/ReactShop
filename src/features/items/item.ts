import { createSlice } from "@reduxjs/toolkit";
import { AppThunk } from "../../app/store";

export interface itemData {
  id: number | undefined;
  quantity: number;
  name: string;
}

export const initialState = {
  id: undefined,
  quantity: 0,
  name: "unknown",
} as itemData;

const ItemSlice = createSlice({
  name: "item",
  initialState,
  reducers: {},
});
