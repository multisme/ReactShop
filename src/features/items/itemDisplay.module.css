.productList{
        list-style: none;
        display: flex;
}

.product{
        width: 150px;
        height: 150px;
}

.centeredContent{
        display: flex;
        align-items: center;
        justify-content: center;
}
