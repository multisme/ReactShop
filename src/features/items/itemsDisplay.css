.itemsList{
        list-style: none;
        display: flex;
        flex-wrap: wrap;
        margin: 0.25rem;
        padding: 0px;
}
