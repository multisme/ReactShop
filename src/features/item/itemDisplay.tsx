import React from "react";
import { useDispatch } from "react-redux";

import { itemData } from "features/item/itemSlice";
import { updateSelectedItem } from "features/showcase/showcaseSlice";

import "features/item/itemDisplay.css";

export const ItemDisplay = (item: itemData) => {
  //change the event type here with the correct one

  const dispatch = useDispatch();

  const selectItem = (e: any) => {
    e.preventDefault();
    dispatch(updateSelectedItem(item));
  };

  return (
    <li
      onClick={selectItem}
      className={"item flex-centered green-border"}
      key={item.id}
    >
      <div className="name">{item.name}</div>
      <div className={"quantity"}>{item.quantity}</div>
    </li>
  );
};
