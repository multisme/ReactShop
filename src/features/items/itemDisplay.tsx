import React from "react";
import {useDispatch} from "react-redux"

import { itemData } from "features/items/item";
import { updateSelectedItem } from "features/showcase/showcaseSlice"

import styles from 'features/items/itemDisplay.module.css';


interface Props{
        items: itemData[]
}


export const ItemListDisplay = ({ items }: Props) => {
        const renderedlist = items.map((item: itemData) => <ItemDisplay {...item}/>);
        return <ul className={styles.productList} id="list">{renderedlist}</ul>;
};

export const ItemDisplay = (item: itemData) => {
//change the event type here with the correct one

const dispatch = useDispatch();

const selectItem = (e: any) => {
        e.preventDefault();
       dispatch( updateSelectedItem(item) );
}

  return (
    <li
        onClick={selectItem}
        className={styles.product} key={item.id}
    >
      <div className="name">{item.name}</div>
      <div className={"quantity"}>{item.quantity}</div>
    </li>
  );
};
