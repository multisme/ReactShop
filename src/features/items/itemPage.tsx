import React from "react";
import { useDispatch, useSelector } from "react-redux";
import {Link} from "react-router-dom";
import {
        useHistory,
        useParams
 } from "react-router-dom";

import { RootState } from "app/rootReducer";
import {itemData} from "features/item/itemSlice";
import {itemPageSelector} from "features/items/itemsSlice";
import {addToCart} from "features/cart/cartSlice";

import { useFormFields } from "utils/utils";

interface productPageParams{
        id: string
}

const ItemPage = ()=>{
        const {id}  = useParams<productPageParams>();
        const [fields, handleFieldChange] = useFormFields({
                quantity: 0
        });
        const dispatch = useDispatch();
        const [item] = useSelector((state) =>itemPageSelector(state, id));
        const history = useHistory();
        if (id == undefined || item == undefined){
                return (
                <h3>Error</h3>
                );
        }
       
        const handleSubmit = (e: any) => {
                e.preventDefault();
                if (fields.quantity.content > 0){
                        dispatch(addToCart({
                        id: parseInt(id),
                        quantity: fields.quantity.content,
                        price: item.price}));
                        history.push("/bill")
                } else {
                        alert("Please choose something before checkout");
                }
        }
        
        const range = [...Array(item.quantity + 1).keys()]
        const availableQuantities = range.map((i) => 
                <option value={i} key={i}>{i}</option>
        )
 
        return (
        <div className="itemPage flex-centered">
                <div className="panel left flex-centered">
                <div className="mainImage">
                <img src="https://teamtijger.nl/wp-content/uploads/2018/06/zwarte-trui-transparant-300x300.png)" />;
                </div>
                </div>
                <div className="panel right">
                        <div className="info">
                                <div className="name">
                                {item.name}
                                </div>
                                <div className="price">
                                {item.price}
                                </div>
                                <div className="details">
                                {item.details}
                                </div>
                        </div>
                <form className="itemForm" onSubmit={handleSubmit}>
                <fieldset className="quantity">
                <select name="quantity" id="quantity" onChange={handleFieldChange}>
                        {availableQuantities}
                </select>
                </fieldset>
                <fieldset className="button">
                <Link to="/">
                <button id="goBack" className="buy" >
                        GO BACK
                </button>
                </Link>
                <button type="submit" id="checkout" className="checkout">
                        CHECKOUT
                </button>
                </fieldset>
                </form>
                </div>
        </div>
        )
}
export default ItemPage;
