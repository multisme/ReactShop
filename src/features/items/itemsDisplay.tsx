import React from "react";

import {useDispatch} from "react-redux"
import {Link} from "react-router-dom"

import { itemData } from "features/item/itemSlice";
import {
        updateSelectedItem,
        removeSelectedItem
} from "features/showcase/showcaseSlice";

import "features/items/itemsDisplay.css";

interface itemListProps{
items: itemData[]
}

interface itemProps{
item: itemData
}

export const ItemListDisplay = ({ items }: itemListProps) => {
        const renderedlist = items.map((item: itemData) =>
                        <ItemDisplay item={item} key={item.id}/>
                        );
        return (
                        <ul className="itemsList" id="list">
                        {renderedlist}
                        </ul>
               );
};

export const ItemDisplay = ({item}: itemProps) => {
        //change the event type here with the correct one

        const dispatch = useDispatch();

        const selectItem = (e: any) => {
                dispatch( updateSelectedItem(item) );
        }

        const unselectItem = (e: any) => {
                dispatch( removeSelectedItem() );
        }

        return (
                        <div className="itemCard"  key={item.id} style={{backgroundImage: "url("  +item.url +")"}}>
                        <Link key={item.id}to={`/products/${item.id}`}>
                        <li
                        onMouseEnter={selectItem}
                        onMouseLeave={unselectItem}
                        className={"flex-centered"}
                        >
                        <div className="picture">
                        </div>
                        <div className={"detailsSmall flex-centered"}>
                        <div className="name">{item.name}</div>
                        <div className={"price"}>{item.price}â‚¬</div>
                        </div>
                        </li>
                        </Link>
                        </div>
               );
};
