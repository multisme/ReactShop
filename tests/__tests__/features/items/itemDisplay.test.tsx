import React from 'react';
import configureStore from 'redux-mock-store';
import { Provider, useDispatch } from 'react-redux';
import {BrowserRouter as Router} from "react-router-dom";
import { mount, shallow } from 'enzyme';

import { ItemDisplay, ItemListDisplay } from "features/items/itemsDisplay";
import { itemData } from "features/item/itemSlice";
import {
        updateSelectedItem,
        removeSelectedItem
} from "features/showcase/showcaseSlice";

const mockStore = configureStore();


describe('itemDisplay', () => {
        let store: any;
        let component: any;

const useDispatchMock = useDispatch as jest.Mock;
                const item: itemData = {id: 3, name: "pierre", quantity: 4};

 
        beforeEach(() => {
                        store = mockStore({ 
                });

                store.dispatch = jest.fn();

                component = mount(
                        <Router>
                        <Provider store={store}>
                                <ItemDisplay item={item} />
                        </Provider>
                        </Router>
                );
        });

        it("component is rendered", () => {
                expect(component).toMatchSnapshot();
        })
        
        it("component dispatch event on hover", () => {
                component.find('li.itemCard').simulate('mouseenter');
                expect(store.dispatch).toHaveBeenCalledTimes(1);
                expect(store.dispatch).toHaveBeenCalledWith(
                        updateSelectedItem(item)
                );
                component.find('li.itemCard').simulate('mouseleave');
                expect(store.dispatch).toHaveBeenCalledTimes(2);
                expect(store.dispatch).toHaveBeenCalledWith(
                        removeSelectedItem()
                );

        })
}) 

describe('ItemListDisplay', () => {
        //Check correct type
        let store: any;
 
        beforeEach(() => {
                store = mockStore({
        });
        });

        it("component is rendered with an empty array", () => {
                const items: itemData[] = [];
                const component= mount(
                        <Router>
                        <Provider store={store}>
                        <ItemListDisplay items={items}/>
                        </Provider>
                        </Router>
                )
                expect(component).toMatchSnapshot();
        });
        it("component is rendered with array of one", () => {
                const items: itemData[] = [
                        {id: 3, "name": "testname", "quantity": 4},
                ];
                const component= mount(
                        <Router>
                        <Provider store={store}>
                        <ItemListDisplay items={items}/>
                        </Provider>
                        </Router>
                )
                expect(component).toMatchSnapshot();
        });
        it("component is rendered with array of some", () => {
                const items: itemData[] = [
                        {id: 3, "name": "testname", "quantity": 4},
                        {id: 4, "name": "postname", "quantity": 8}
                ];
                const component= mount(
                        <Router>
                        <Provider store={store}>
                        <ItemListDisplay items={items}/>
                        </Provider>
                        </Router>
                );
                expect(component).toMatchSnapshot();
        })
})
