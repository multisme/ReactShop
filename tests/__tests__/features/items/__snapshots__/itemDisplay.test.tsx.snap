// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ItemListDisplay component is rendered with an empty array 1`] = `
<BrowserRouter>
  <Router
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "createHref": [Function],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "pathname": "/",
          "search": "",
          "state": undefined,
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <Provider
      store={
        Object {
          "clearActions": [Function],
          "dispatch": [Function],
          "getActions": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        }
      }
    >
      <ItemListDisplay
        items={Array []}
      >
        <ul
          className="itemsList"
          id="list"
        />
      </ItemListDisplay>
    </Provider>
  </Router>
</BrowserRouter>
`;

exports[`ItemListDisplay component is rendered with array of one 1`] = `
<BrowserRouter>
  <Router
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "createHref": [Function],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "pathname": "/",
          "search": "",
          "state": undefined,
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <Provider
      store={
        Object {
          "clearActions": [Function],
          "dispatch": [Function],
          "getActions": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        }
      }
    >
      <ItemListDisplay
        items={
          Array [
            Object {
              "id": 3,
              "name": "testname",
              "price": 3,
              "quantity": 4,
            },
          ]
        }
      >
        <ul
          className="itemsList"
          id="list"
        >
          <ItemDisplay
            item={
              Object {
                "id": 3,
                "name": "testname",
                "price": 3,
                "quantity": 4,
              }
            }
            key="3"
          >
            <div
              className="itemCard"
              key="3"
              style={
                Object {
                  "backgroundImage": "url(undefined)",
                }
              }
            >
              <Link
                key="3"
                to="/products/3"
              >
                <LinkAnchor
                  href="/products/3"
                  navigate={[Function]}
                >
                  <a
                    href="/products/3"
                    onClick={[Function]}
                  >
                    <li
                      className="flex-centered"
                      onMouseEnter={[Function]}
                      onMouseLeave={[Function]}
                    >
                      <div
                        className="picture"
                      />
                      <div
                        className="detailsSmall flex-centered"
                      >
                        <div
                          className="name"
                        >
                          testname
                        </div>
                        <div
                          className="price"
                        >
                          3
                          €
                        </div>
                      </div>
                    </li>
                  </a>
                </LinkAnchor>
              </Link>
            </div>
          </ItemDisplay>
        </ul>
      </ItemListDisplay>
    </Provider>
  </Router>
</BrowserRouter>
`;

exports[`ItemListDisplay component is rendered with array of some 1`] = `
<BrowserRouter>
  <Router
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "createHref": [Function],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "pathname": "/",
          "search": "",
          "state": undefined,
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <Provider
      store={
        Object {
          "clearActions": [Function],
          "dispatch": [Function],
          "getActions": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        }
      }
    >
      <ItemListDisplay
        items={
          Array [
            Object {
              "id": 3,
              "name": "testname",
              "price": 3,
              "quantity": 4,
            },
            Object {
              "id": 4,
              "name": "postname",
              "price": 2,
              "quantity": 8,
            },
          ]
        }
      >
        <ul
          className="itemsList"
          id="list"
        >
          <ItemDisplay
            item={
              Object {
                "id": 3,
                "name": "testname",
                "price": 3,
                "quantity": 4,
              }
            }
            key="3"
          >
            <div
              className="itemCard"
              key="3"
              style={
                Object {
                  "backgroundImage": "url(undefined)",
                }
              }
            >
              <Link
                key="3"
                to="/products/3"
              >
                <LinkAnchor
                  href="/products/3"
                  navigate={[Function]}
                >
                  <a
                    href="/products/3"
                    onClick={[Function]}
                  >
                    <li
                      className="flex-centered"
                      onMouseEnter={[Function]}
                      onMouseLeave={[Function]}
                    >
                      <div
                        className="picture"
                      />
                      <div
                        className="detailsSmall flex-centered"
                      >
                        <div
                          className="name"
                        >
                          testname
                        </div>
                        <div
                          className="price"
                        >
                          3
                          €
                        </div>
                      </div>
                    </li>
                  </a>
                </LinkAnchor>
              </Link>
            </div>
          </ItemDisplay>
          <ItemDisplay
            item={
              Object {
                "id": 4,
                "name": "postname",
                "price": 2,
                "quantity": 8,
              }
            }
            key="4"
          >
            <div
              className="itemCard"
              key="4"
              style={
                Object {
                  "backgroundImage": "url(undefined)",
                }
              }
            >
              <Link
                key="4"
                to="/products/4"
              >
                <LinkAnchor
                  href="/products/4"
                  navigate={[Function]}
                >
                  <a
                    href="/products/4"
                    onClick={[Function]}
                  >
                    <li
                      className="flex-centered"
                      onMouseEnter={[Function]}
                      onMouseLeave={[Function]}
                    >
                      <div
                        className="picture"
                      />
                      <div
                        className="detailsSmall flex-centered"
                      >
                        <div
                          className="name"
                        >
                          postname
                        </div>
                        <div
                          className="price"
                        >
                          2
                          €
                        </div>
                      </div>
                    </li>
                  </a>
                </LinkAnchor>
              </Link>
            </div>
          </ItemDisplay>
        </ul>
      </ItemListDisplay>
    </Provider>
  </Router>
</BrowserRouter>
`;

exports[`itemDisplay component is rendered 1`] = `
<BrowserRouter>
  <Router
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "createHref": [Function],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "pathname": "/",
          "search": "",
          "state": undefined,
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <Provider
      store={
        Object {
          "clearActions": [Function],
          "dispatch": [MockFunction],
          "getActions": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        }
      }
    >
      <ItemDisplay
        item={
          Object {
            "id": 3,
            "name": "pierre",
            "price": 3,
            "quantity": 4,
          }
        }
      >
        <div
          className="itemCard"
          key="3"
          style={
            Object {
              "backgroundImage": "url(undefined)",
            }
          }
        >
          <Link
            key="3"
            to="/products/3"
          >
            <LinkAnchor
              href="/products/3"
              navigate={[Function]}
            >
              <a
                href="/products/3"
                onClick={[Function]}
              >
                <li
                  className="flex-centered"
                  onMouseEnter={[Function]}
                  onMouseLeave={[Function]}
                >
                  <div
                    className="picture"
                  />
                  <div
                    className="detailsSmall flex-centered"
                  >
                    <div
                      className="name"
                    >
                      pierre
                    </div>
                    <div
                      className="price"
                    >
                      3
                      €
                    </div>
                  </div>
                </li>
              </a>
            </LinkAnchor>
          </Link>
        </div>
      </ItemDisplay>
    </Provider>
  </Router>
</BrowserRouter>
`;
